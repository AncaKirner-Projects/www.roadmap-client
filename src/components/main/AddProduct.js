import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
// import MaterialIcon from 'material-icons-react';
import IconButton from '@material-ui/core/IconButton';
import { ValidatorForm, TextValidator, SelectValidator } from 'react-material-ui-form-validator';
import styles from '../layers/styles/addProductStyles';
import { addProduct, deleteStatusMessage } from '../../store/actions/productActions';
import AlertDialog from './AlertDialog';
import ActionTypes from '../../store/actions/actionTypes';

class AddProduct extends Component {
  state = {
    open: false,
    openAlertDialog: false,
    formData: {
      category: '',
      name: '',
      description: '',
      price: '',
      totalProducts: ''
    }
  };

  componentDidMount() {
    ValidatorForm.addValidationRule('isNumber', value => !isNaN(parseInt(value, 0)));
  }

  handleClickOpen = () => {
    this.setState({ open: true });
  }

  handleClose = () => {
    this.setState({
      open: false,
      formData: {
        category: 0
      }
    });
  }

  handleAlertDialogClose = () => {
    this.props.deleteStatusMessage(ActionTypes.ADD_PRODUCT);
    this.setState({ openAlertDialog: false });
  }

  handleChange = name => (e) => {
    const { formData } = this.state;
    formData[name] = e.target.value;
    this.setState({ formData });
  }

  handleCreateProduct = () => {
    const product = {
      category_id: parseInt(this.state.formData.category, 0),
      prod_name: this.state.formData.name,
      prod_description: this.state.formData.description,
      price: parseFloat(this.state.formData.price).toFixed(2),
      prod_number: parseInt(this.state.formData.totalProducts, 0)
    };
    this.props.addProduct(product);
    this.setState({ open: false, openAlertDialog: true });
  }

  render() {
    const { categories, status } = this.props;
    const { formData } = this.state;
    const addProductStatus = status && status.find(elem => elem.type === ActionTypes.ADD_PRODUCT);

    return (
      <React.Fragment>
        <IconButton color="inherit" onClick={this.handleClickOpen}>
          {/* <MaterialIcon icon="playlist_add" color="white" size={35} /> */}
        </IconButton>
        <Dialog
          open={this.state.open}
          onClose={this.handleClose}
          aria-labelledby="form-dialog-title"
        >
          <ValidatorForm
            ref="form"
            onSubmit={this.handleCreateProduct}
            onError={errors => console.log(errors)}
          >
            <DialogTitle id="form-dialog-title">Add product</DialogTitle>
            <DialogContent>
              <SelectValidator
                id="standard-select-currency-native"
                autoFocus
                select
                label="Select category"
                className={styles.textField}
                value={formData.category}
                onChange={this.handleChange('category')}
                SelectProps={{
                  native: true,
                  MenuProps: {
                    className: styles.menu
                  }
                }}
                margin="normal"
                fullWidth
                name="category"
                validators={['required']}
                errorMessages={['This field is required']}
              >
                <option key={0} value="Please select">
                  - Please select -
                </option>
                {categories && categories.map(option => (
                  <option key={option.id} value={option.id}>
                    {option.name}
                  </option>
                ))}
              </SelectValidator>
              <TextValidator
                margin="dense"
                id="name"
                label="Product Name"
                fullWidth
                name="name"
                value={formData.name}
                onChange={this.handleChange('name')}
                validators={['required']}
                errorMessages={['This field is required']}
              />
              <TextValidator
                id="standard-textarea"
                label="Product Description"
                multiline
                margin="normal"
                fullWidth
                name="description"
                value={formData.description}
                onChange={this.handleChange('description')}
                validators={['required']}
                errorMessages={['This field is required']}
              />
              <TextValidator
                margin="dense"
                id="name"
                label="Price"
                fullWidth
                name="price"
                value={formData.price}
                onChange={this.handleChange('price')}
                validators={['required', 'isNumber']}
                errorMessages={['This field is required', 'Price is not valid']}
              />
              <TextValidator
                margin="dense"
                id="name"
                label="Number of products"
                fullWidth
                name="totalProducts"
                value={formData.totalProducts}
                onChange={this.handleChange('totalProducts')}
                validators={['required', 'isNumber']}
                errorMessages={['This field is required', 'Number of products is not valid']}
              />
            </DialogContent>
            <DialogActions>
              <Button
                variant="contained"
                type="submit"
                color="primary"
              >
                Create
              </Button>
            </DialogActions>
          </ValidatorForm>
        </Dialog>
        <AlertDialog
          open={this.state.openAlertDialog}
          handleClose={this.handleAlertDialogClose}
          title="Request Status"
          text={addProductStatus && addProductStatus.message}
        />
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => ({
  categories: state.categories.list,
  status: state.products.status
});

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({
    addProduct,
    deleteStatusMessage
  }, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(AddProduct);
