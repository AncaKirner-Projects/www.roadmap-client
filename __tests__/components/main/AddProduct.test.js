import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import AddProduct from '../../../src/components/main/AddProduct';
import ActionTypes from '../../../src/store/actions/actionTypes';

const mockStore = configureMockStore();

describe('<AddProduct />', () => {
  const store = mockStore({
    categories: {
      list: [
        { id: 1, name: 'Categ1' },
        { id: 2, name: 'Categ2' }
      ]
    },
    products: {
      status: [
        {
          type: ActionTypes.ADD_PRODUCT,
          message: 'Test message'
        }
      ]
    }
  });

  const props = {
    addProduct: jest.fn(),
    deleteStatusMessage: jest.fn()
  };

  it('renders an element', () => {
    const wrapper = shallow(<AddProduct {...props} store={store} />);
    const component = wrapper.dive();

    expect(toJson(component)).toMatchSnapshot();
  });
});
