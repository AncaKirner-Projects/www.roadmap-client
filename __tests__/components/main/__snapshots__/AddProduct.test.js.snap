// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddProduct /> renders an element 1`] = `
<Fragment>
  <WithStyles(IconButton)
    color="inherit"
    onClick={[Function]}
  />
  <WithStyles(Dialog)
    aria-labelledby="form-dialog-title"
    onClose={[Function]}
    open={false}
  >
    <ValidatorForm
      debounceTime={0}
      onError={[Function]}
      onSubmit={[Function]}
    >
      <WithStyles(DialogTitle)
        id="form-dialog-title"
      >
        Add product
      </WithStyles(DialogTitle)>
      <WithStyles(DialogContent)>
        <SelectValidator
          SelectProps={
            Object {
              "MenuProps": Object {
                "className": undefined,
              },
              "native": true,
            }
          }
          autoFocus={true}
          errorMessages={
            Array [
              "This field is required",
            ]
          }
          fullWidth={true}
          id="standard-select-currency-native"
          label="Select category"
          margin="normal"
          name="category"
          onChange={[Function]}
          select={true}
          validatorListener={[Function]}
          validators={
            Array [
              "required",
            ]
          }
          value=""
        >
          <option
            key="0"
            value="Please select"
          >
            - Please select -
          </option>
          <option
            key="1"
            value={1}
          >
            Categ1
          </option>
          <option
            key="2"
            value={2}
          >
            Categ2
          </option>
        </SelectValidator>
        <TextValidator
          errorMessages={
            Array [
              "This field is required",
            ]
          }
          fullWidth={true}
          id="name"
          label="Product Name"
          margin="dense"
          name="name"
          onChange={[Function]}
          validatorListener={[Function]}
          validators={
            Array [
              "required",
            ]
          }
          value=""
        />
        <TextValidator
          errorMessages={
            Array [
              "This field is required",
            ]
          }
          fullWidth={true}
          id="standard-textarea"
          label="Product Description"
          margin="normal"
          multiline={true}
          name="description"
          onChange={[Function]}
          validatorListener={[Function]}
          validators={
            Array [
              "required",
            ]
          }
          value=""
        />
        <TextValidator
          errorMessages={
            Array [
              "This field is required",
              "Price is not valid",
            ]
          }
          fullWidth={true}
          id="name"
          label="Price"
          margin="dense"
          name="price"
          onChange={[Function]}
          validatorListener={[Function]}
          validators={
            Array [
              "required",
              "isNumber",
            ]
          }
          value=""
        />
        <TextValidator
          errorMessages={
            Array [
              "This field is required",
              "Number of products is not valid",
            ]
          }
          fullWidth={true}
          id="name"
          label="Number of products"
          margin="dense"
          name="totalProducts"
          onChange={[Function]}
          validatorListener={[Function]}
          validators={
            Array [
              "required",
              "isNumber",
            ]
          }
          value=""
        />
      </WithStyles(DialogContent)>
      <WithStyles(DialogActions)>
        <WithStyles(Button)
          color="primary"
          type="submit"
          variant="contained"
        >
          Create
        </WithStyles(Button)>
      </WithStyles(DialogActions)>
    </ValidatorForm>
  </WithStyles(Dialog)>
  <AlertDialog
    handleClose={[Function]}
    open={false}
    text="Test message"
    title="Request Status"
  />
</Fragment>
`;
